### PR TITLE
Fix idempotent issue when using lvol with %VG or %PVS size options and VG is fully allocated

### LINKED ISSUE #290: lvol: Ansible on 2nd and later runs expanding the Logical Volume size by doubling it, rather than locking it at 50% of the VG
##### SUMMARY

Issue is caused by Ansible on 2nd and later runs expanding the Logical Volume size by doubling it, rather than locking it at 50% of the VG.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

lvol

##### ANSIBLE VERSION

```
ansible 2.9.7
  config file = None
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /root/.local/lib/python2.7/site-packages/ansible
  executable location = /root/bin/ansible
  python version = 2.7.5 (default, Nov 27 2019, 09:57:45) [GCC 4.8.5 20150623 (Red Hat 4.8.5-39.0.1)]
```

##### CONFIGURATION

```paste below

```

##### OS / ENVIRONMENT

OS version: OracleLinux 7.x

##### STEPS TO REPRODUCE

```
- name: 'Ensure partitions exist'
  parted:
    align: 'optimal'
    device: '{{ item }}'
    label: 'gpt'
    name: 'nvme'
    number: '1'
    part_end: '100%'
    part_start: '0%'
    state: 'present'
  loop:
     - '/dev/nvme0n1'
     - '/dev/nvme1n1'
- name: 'Ensure vg_nvme exists'
  lvg:
    pvs: /dev/nvme0n1p1,/dev/nvme1n1p1
    state: 'present'
    vg: 'vg_nvme'
- name: 'Ensure lv_nvme exists'
  lvol:
    active: 'yes'
    lv: 'lv_nvme'
    opts: '--type raid1'
    resizefs: 'yes'
    shrink: 'yes'
    size: '50%VG'
    state: 'present'
    vg: 'vg_nvme'
- name: 'Ensure filesystem exists'
  filesystem:
    dev: '/dev/mapper/vg_nvme-lv_nvme'
    fstype: 'xfs'
- name: 'Ensure filesystem mounted'
  mount:
    backup: 'yes'
    dump: '1'
    fstype: 'xfs'
    opts: 'defaults'
    passno: '2'
    path: '/nvme'
    src: '/dev/mapper/vg_nvme-lv_nvme'
    state: 'mounted'


```

##### ACTUAL RESULTS

```
first run:
[root@vmdenseio216-test ~]# df -hP /nvme
Filesystem Size Used Avail Use% Mounted on
/dev/mapper/vg_nvme-lv_nvme 3.0T 33M 3.0T 1% /nvme
[root@vmdenseio216-test ~]# vgs
VG #PV #LV #SN Attr VSize VFree
vg_nvme 2 1 0 wz--n- 11.64t 5.82t
[root@vmdenseio216-test ~]# lvs
LV VG Attr LSize Pool Origin Data% Meta% Move Log Cpy%Sync Convert
lv_nvme vg_nvme rwi-aor--- 2.91t 0.52
[root@vmdenseio216-test ~]# pvs
PV VG Fmt Attr PSize PFree
/dev/nvme0n1p1 vg_nvme lvm2 a-- 5.82t 2.91t
/dev/nvme1n1p1 vg_nvme lvm2 a-- 5.82t 2.91t
[root@vmdenseio216-test ~]#

Second run:
[root@vmdenseio216-test ~]# df -hT /nvme
Filesystem Type Size Used Avail Use% Mounted on
/dev/mapper/vg_nvme-lv_nvme xfs 5.9T 33M 5.9T 1% /nvme
[root@vmdenseio216-test ~]# vgs
VG #PV #LV #SN Attr VSize VFree
vg_nvme 2 1 0 wz--n- 11.64t 0
[root@vmdenseio216-test ~]# lvs
LV VG Attr LSize Pool Origin Data% Meta% Move Log Cpy%Sync Convert
lv_nvme vg_nvme rwi-aor--- 5.82t 0.51
[root@vmdenseio216-test ~]# pvs
PV VG Fmt Attr PSize PFree
/dev/nvme0n1p1 vg_nvme lvm2 a-- 5.82t 0
/dev/nvme1n1p1 vg_nvme lvm2 a-- 5.82t 0
[root@vmdenseio216-test ~]#

```

### PR DESCRIPTION
**Pull request by JuddTracy-DAS**
##### SUMMARY
Changed the behavior when using %VG or %PVS to make the size_requested an even modulus with the VGs physical extents by rounding down.  This makes the usage of %VG or %PVS idempotent when the calculated size_requested does not end on a physical extent boundary and the VG is fully allocated.

Fixes #290.

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
lvol

##### ADDITIONAL INFORMATION

The VG looks like this:

> vgs --noheadings --nosuffix -o vg_name,size,free,vg_extent_size --units 'm' --separator ';' mysql_vg
>
> mysql_vg;3060.00;0;4.00

Based on the logic of the module the size requested looks like this:

> size_requested = size_percent * this_vg['size'] / 100
> 2295.0 = 25 * 3060.00 / 100
>

When divided by the extent size we can see it is not whole number:
> size_requested / 4 = 573.75

This is not an issue unless the VG is fully allocated.  When that is the case you get the following error:

> fatal: [mysql]: FAILED! => {"changed": false, "msg": "Logical Volume var_lib_mysql2 could not be extended. Not enough free space left (1.0m required / 0.0m available)"}

Example playbook:
```
    tasks:
    - lvol:
        vg: 'mysql_vg'
        lv: 'var_lib_mysql2'
        size: '25%VG'
        shrink: no
        state: 'present'
```

### CODE REVIEW

File: changelogs/fragments/229_lvol_percentage_fix.yml
Comment (felixfontein) [inline changelogs/fragments/229_lvol_percentage_fix.yml:2]:
```suggestion
 - lvol - Fix idempotent issue when using lvol with %VG or %PVS size options and VG is fully allocated (https://github.com/ansible-collections/community.general/pull/229).
```

**Comment by felixfontein** [comment]
@MorrisA @abulimov @bcoca @d-little @flynn1973 @gforster @jhoekx @kairoaraujo @marvin-sinister @mator @molekuul @ramooncamacho @wtcross since you are maintainers for this module, could you please check whether the code change is correct?

File: changelogs/fragments/229_lvol_percentage_fix.yml
Comment (JuddTracy-DAS) [reply changelogs/fragments/229_lvol_percentage_fix.yml]: Done

**Comment by Andersson007** [comment]
@JuddTracy-DAS any updates on this?

**Comment by Andersson007** [comment]
cc @jhoekx @abulimov
@JuddTracy-DAS could you also please email the authors directly? There are their emails in the module's header

**Comment by JuddTracy-DAS** [comment]
I merged the suggestions from felixfontein, is there something else you need done?

**Comment by Andersson007** [comment]
@JuddTracy-DAS not on my side but i don't use the module, better to have approvals from authors or users. Did you email them?

**Review by Andersson007 — APPROVED** [review]

**Comment by Andersson007** [comment]
@JuddTracy-DAS if we merge this, could we add you to the module's maintainers? so you will receive notifications about related events (issues/PRs) and your shipit will be counted by the bot for fixes/minor-changes automerge

**Comment by felixfontein** [comment]
@MorrisA @abulimov @bcoca @d-little @flynn1973 @gforster @jhoekx @kairoaraujo @marvin-sinister @mator @molekuul @ramooncamacho @wtcross as the module maintainers, could you please take a look at this change?

**Comment by Andersson007** [comment]
@vburamdo @mator @danielmotaleite does anybody use this module? That needs to add only one line to check

**Comment by BenoitKnecht** [comment]
@Andersson007 I'm affected by this issue, and I can confirm that this PR fixes it.

Before applying this fix, running a playbook with `lvol` using `size=8%VG`:

1. First Ansible run creates the logical volume (19201523712 bytes in my case);
2. Second Ansible run reports the `lvol` result as `changed`, and actually extends the logical volume by a few bytes (for a total of 19205718016 bytes);
3. Thrid Ansible run onward reports no changes and doesn't resize the logical volume.

After applying the fix:

1. First Ansible run creates the logical volume (19201523712 bytes in my case);
2. Second Ansible run onward reports no changes and doesn't resize the logical volume.

which is the expected behavior.

**Review by felixfontein — APPROVED** [review]
If nobody complains, I'll merge this beginning of next week.

**Comment by Andersson007** [comment]
@BenoitKnecht great, thanks for the feedback!

**Comment by felixfontein** [comment]
@JuddTracy-DAS thanks for fixing this!
@Andersson007 @BenoitKnecht thanks for reviewing and testing this!

**Comment by BenoitKnecht** [comment]
Is this going to be automatically backported to [ansible/ansible](https://github.com/ansible/ansible), or should a PR be manually submitted there?

**Comment by felixfontein** [comment]
This needs to be manually backported. If you want to do that, you need to create a PR there for the stable-2.9 branch, create exactly the same changes as here (with paths adjusted), and reference this PR.

**Comment by BenoitKnecht** [comment]
Thanks, done.

**Pull request merged**
